### PR TITLE
ci: add sha256sum and version.txt

### PR DESCRIPTION
**Pull request by Itsusinn**

### CODE REVIEW

**Comment by Itsusinn** [comment]
#596 #597

**Comment by Itsusinn** [comment]
@greenhat616 If you have any alternative requirement，please tell me

**Comment by greenhat616** [comment]
Maybe we could use the `cargo set-version` from [cargo-edit](https://crates.io/crates/cargo-edit) to make the consistent with the `version.txt`?
And here is the spec of semver: <https://semver.org/#backusnaur-form-grammar-for-valid-semver-versions>
So that `x.y.z-tag+buildinfo` is vaild for semver.

If use are inconsistant with the version.txt. The updater should create a independent lockfile to fix version by file chksum, which is not implemented yet now.

Here is the mihomo alpha channel output:
```
Mihomo Meta alpha-3676d1b windows amd64 with go1.23.1 Wed Sep 18 14:44:10 UTC 2024
Use tags: with_gvisor
```

It is consistent with the content of `version.txt`.
For manifest v1, we parsed the version, using it compare with the remote `version.txt`, and do the update logic.

**Comment by Itsusinn** [comment]
> we might use the `cargo set-version` from [cargo-edit](https://crates.io/crates/cargo-edit) to make the consistent with the `version.txt`?

Doing so will crate tons of commits by bot（kinda of spam i have to say）.

So no.

> If the program version are inconsistant with the `version.txt`. The updater are supposed to create a independent lockfile to fix version by file chksum, which is not implemented yet now.

Is the `program version` you pointed out here meaning cli's `clash-rs -v` output？

If so，that's doable.

**Comment by greenhat616** [comment]
> we might use the `cargo set-version` from [cargo-edit](https://crates.io/crates/cargo-edit) to make the consistent with the `version.txt`?

I think we do not need to commit the changes after modify the version for alpha build, because we don't release a real alpha tag in github like other well-known repos do.

> > If the program version are inconsistant with the `version.txt`. The updater are supposed to create a independent lockfile to fix version by file chksum, which is not implemented yet now.
> 
> Is the `program version` you pointed out here meaning cli's `clash-rs -v` output？
> 
> If so，that's possible to do.

Yes

**Comment by Itsusinn** [comment]
> I think we do not need to commit the changes after modify the version for alpha build, because we don't release a real alpha tag in github like other well-known repos do.

Very good thinking!

Now new version (.txt) format in my mind: `0.x.y-alpha+sha.000000`

**Comment by greenhat616** [comment]
> > I think we do not need to commit the changes after modify the version for alpha build, because we don't release a real alpha tag in github like other well-known repos do.
> 
> Very good thinking!
> 
> Now new version (.txt) format in my mind: `0.x.y-alpha+sha.0000000`

Yeah. That's enough.

**Comment by ibigbug** [comment]
Yeah let's also change the cli version output for the nightly release to make it aligned

**Comment by greenhat616** [comment]
> > I think we do not need to commit the changes after modify the version for alpha build, because we don't release a real alpha tag in github like other well-known repos do.
> 
> Very good thinking! But it still makes action more complex, we have to do these things **in every target**:
> 
>     * Parse current version from `Cargo.toml`
> 
>     * Get short SHA from `git`
> 
>     * Install `cargo-edit` by `cargo install` (another few minutes)
> 
>     * Join vars, and `cargo set new-version`
> 
> 
> I would prefer add a environment called `VERSION_OVERRIDE` And in cli process:
> 
> ```rust
> let version = match env!(VERSION_OVERRIDE) {
>     Some(v) -> v
>     None -> env!(CARGO_PKG_VERSION)
> }
> ```
> 
> Now new version (.txt) format in my mind: `0.x.y-alpha+sha.0000000`

Yeah. If we use the `cargo-edit`, we'd better use [cargo-binstall](https://github.com/cargo-bins/cargo-binstall?tab=readme-ov-file) to install it.

And the build envs var is a good way too.

File: clash/src/main.rs [hunk 1]
```diff
@@ -31,13 +31,27 @@ struct Cli {
...
+    version: bool,
     #[clap(short, long, help = "Additinally log to file")]
     log_file: Option<String>,
 }
 
 fn main() {
     let cli = Cli::parse();
+
+    if cli.version {
+        println!("{}", env!("CLASH_VERSION_OVERRIDE"));
```
Comment (greenhat616) [inline clash/src/main.rs]: could follow the current style like below?

```shell
clash-rs 0.3.2
```

**Review by greenhat616 — APPROVED** [review]
LGTM

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -377,13 +382,20 @@ jobs:
           mv ./clash-rs/clash-x86_64-unknown-linux-musl ./clash-rs/clash-amd64
           mv ./clash-rs/clash-aarch64-unknown-linux-musl ./clash-rs/clash-arm64
 
+      - name: Log in to Container Registry
```
Comment (ibigbug) [inline .github/workflows/ci.yml:385]: is this intentional?

File: clash/build.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+#![feature(let_chains)]
+fn main() {
+    println!("cargo:rerun-if-env-changed=CARGO_PKG_VERSION");
```
Comment (ibigbug) [inline clash/build.rs:3]: this wouldn't rebuild on master when only ref changes but not pkg version right?

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Itsusinn) [reply .github/workflows/ci.yml:385]: Yes，I about to refactor docker workflow，but failed. Probably another pr

File: clash/build.rs (hunk 3 above)
Comment (Itsusinn) [reply clash/build.rs:3]: right, But for now it only cares package version. If someday we have extra requirements, it's not late to change it then.

Comment (ibigbug) [reply clash/build.rs:3]: But you'd want to rebuild for every master commit when version doesn't change?

Comment (Itsusinn) [reply clash/build.rs:3]: Will rebuild happen when actions are traiggered? 

I am a little not sure for now, what should I change this into?

Comment (ibigbug) [reply clash/build.rs:3]: Actually it should be fine. Any code change on master should trigger a rebuild anyways.

**Review by ibigbug — APPROVED** [review]

**Comment by Itsusinn** [comment]
It seems broken......

**Pull request merged**